### PR TITLE
Feature: shim: Complete the metrics collection for cgroup v1

### PR DESCRIPTION
**Pull request by zzzzzzzzzy9**
Complete the metrics collection for cgroup v1 so that containerd can obtain complete metrics information

### CODE REVIEW

**Comment by zzzzzzzzzy9** [comment]
@mxpv

**Comment by zzzzzzzzzy9** [comment]
@Burning1020

File: crates/shim/src/cgroup.rs [hunk 1]
```diff
@@ -102,15 +115,194 @@ fn write_process_oom_score(pid: u32, score: i64) -> Result<()> {
...
+                for line in stat.lines() {
+                    let parts = line.split(' ').collect::<Vec<&str>>();
+                    if parts.len() != 2 {
+                        Err(Error::Other(format!("invalid cpu stat line: {}", line)))?;
+                    }
+
+                    // https://github.com/opencontainers/runc/blob/dbe8434359ca35af1c1e10df42b1f4391c1e1010/libcontainer/cgroups/fs2/cpu.go#L70
+                    match parts[0] {
+                        "nr_periods" => {
+                            throttle.set_periods(parts[1].parse::<u64>().unwrap());
```
Comment (Burning1020) [inline crates/shim/src/cgroup.rs:150]: Could you please use `unwrap_or_default()` to avoiding panic directly?

File: crates/shim/src/cgroup.rs [hunk 2]
```diff
@@ -184,6 +369,134 @@ pub fn collect_metrics(pid: u32) -> Result<Metrics> {
...
+                        if *s[0] == device_key.major.to_string()
+                            && *s[1] == device_key.minor.to_string()
+                        {
+                            return Ok(s[3].to_string());
+                        }
+                    }
+                    // if we don't find the device in /proc/partitions, just return a nil string.
+                    Ok("".to_string())
+                }
+                Err(e) => Err(Error::Other(
```
Comment (Burning1020) [inline crates/shim/src/cgroup.rs:401]: we have `other!()`

**Comment by Mossaka** [comment]
I think it's great to have a complete cgroupv1 stats collection, but given https://github.com/containerd/rust-extensions/issues/258 suggests that the complex logic of cgroups stats should not reside in the shim crate, I would recommend that we should take a look at if [libcgroups](https://github.com/containers/youki/tree/main/crates/libcgroups) could be of use.

In particular, `libcgroups` maintained by youki has cgroups v1 and v2 `StatsProvider` ([source](https://github.com/containers/youki/blob/main/crates/libcgroups/src/v1/cpu.rs#L59-L87)), so we don't need to duplicate the effort in this repo. 

I am fully aware that we are using [cgroups-rs](https://github.com/kata-containers/cgroups-rs), so what I am proposing here is to evaluate whether or not `libcgroups` satisfies our requirements and can replace `cgroups-rs`.

CC @utam0k

**Comment by utam0k** [comment]
I'm happy to collaborate with you if possible!

File: crates/shim/src/cgroup.rs (hunk 1 above)
Comment (zzzzzzzzzy9) [reply crates/shim/src/cgroup.rs]: Done.

File: crates/shim/src/cgroup.rs (hunk 2 above)
Comment (zzzzzzzzzy9) [reply crates/shim/src/cgroup.rs]: Done.

**Comment by mxpv** [comment]
Hey @Mossaka, can you give a bit more context on `cgroups-rs` vs `libcgroups` ?

**Comment by mxpv** [comment]
Thanks for your work on this! Since the long-term plan is to remove cgroups from the shim crate (see [#258](https://github.com/containerd/rust-extensions/issues/258)), I’m going to close this PR.

**Pull request closed without merging**